src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -2067,7 +2069,28 @@ because xbuild doesn't support framework reference assemblies.
 </Target>
 	
 <Target Name="_LinkAssemblies"
-  DependsOnTargets="_ResolveAssemblies;_CreatePackageWorkspace;_LinkAssembliesNoShrink;_LinkAssembliesShrink" />
+  DependsOnTargets="_ResolveAssemblies;_CreatePackageWorkspace;_GenerateJniMarshalMethods;_LinkAssembliesNoShrink;_LinkAssembliesShrink" />
+
+<Target Name="_GenerateJniMarshalMethods"
+    Condition="'$(AndroidGenerateJniMarshalMethods)' == 'True' And '$(AndroidLinkMode)' != 'None'"
+    DependsOnTargets="_GetReferenceAssemblyPaths;_SetLatestTargetFrameworkVersion"
+    Inputs="@(ResolvedUserAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')"
+    Outputs="$(_AndroidJniMarshalMethodsFlag)">
+  <ItemGroup>
+    <_JniFrameworkAssembly Include="Mono.Android.dll" />
+    <_JniFrameworkAssembly Include="OpenTK-1.0.dll" />
+    <_JniFrameworkAssembly Include="OpenTK.dll" />
+    <_JniFrameworkAssembly Include="Xamarin.Android.NUnitLite.dll" />
+  </ItemGroup>
+  <ItemGroup>
+    <_AssembliesToProcess Include="@(ResolvedUserAssemblies)" />
+    <_AssembliesToProcess Include="@(ResolvedAssemblies)" Condition=" '%(Filename)' != '' And '@(_JniFrameworkAssembly)' != '' " />
+  </ItemGroup>
+  <Exec
+       Command="MONO_PATH=&quot;$(_XATargetFrameworkDirectories)&quot; &quot;$(MonoAndroidBinDirectory)\mono&quot; &quot;$(MonoAndroidToolsDirectory)\jnimarshalmethod-gen.exe&quot; --jvm=&quot;$(JdkJvmPath)&quot; @(_AssembliesToProcess->'&quot;$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)&quot;', ' ')"
```

Comment:
@jonpryor what is our preference there? using `Exec` or writing a `Task` to do this job? 
If we add a test... we can add a unit test.....

Reply:
@radekdoulik that is a good point. How will this work on windows? mono is not installed as part of windows.

Reply:
It might be a good idea to use `mono --debug` here, to get filename & line number information from exception messages.

Reply:
@dellis1972: We'll need to update the Xamarin.Android install so that we *do* install `mono` on Windows (and macOS, for that matter).

@radekdoulik: Either this PR or another will need to update `src/mono-runtimes` so that `mono.exe` for Windows is included into the bundle.  A Darwin `mono` is already being included.

@radekdoulik: This may in turn require more "flexibility": iirc Mono on Windows has both a `mono.exe` and a `monow.exe`, which differ only on the Windows Subsystems that they target.  Executing `mono.exe` may result in a "console window" "flashing" on the user's screen, which is *really* annoying.  `monow.exe` *won't* cause such flashing, iirc.

Or we might want to use a dedicated Task, as @dellis1972 suggests, either so that we can use `monow` on Windows, or so that we can use a specific `System.Diagnostics.Process` which sets `ProcessStartInfo.CreateNoWindow=true` and `ProcessStartInfo.WindowStyle=ProcessWindowStyle.Hidden`, as is done elsewhere.